buildSrc/src/main/java/fmlbuild/InstallProductionClientTask.java

```diff
@@ -0,0 +1,325 @@
+     * A JVM compatible Arg-File that only contains the name of the main-class used by Vanilla.
+     */
+    @OutputFile
+    public abstract RegularFileProperty getVanillaMainClassArgFile();
+
+    /**
+     * A File that contains one line per program argument passed to Vanilla. Encoded as UTF-8, unlike
+     * the normal JVM arg-files.
+     */
+    @OutputFile
+    public abstract RegularFileProperty getVanillaProgramArgFile();
+
+    /**
+     * Points to a JVM compatible Arg-File, which contains the JVM parameters needed to launch NeoForge.
+     * This is additive to {@link #getVanillaJvmArgFile()}.
+     */
+    @OutputFile
+    public abstract RegularFileProperty getNeoForgeJvmArgFile();
+
+    /**
+     * A JVM compatible Arg-File that contains the class-name used by NeoForge for launching.
+     * This is mutually exclusive with {@link #getVanillaMainClassArgFile()}.
+     */
+    @OutputFile
+    public abstract RegularFileProperty getNeoForgeMainClassArgFile();
+
+    /**
+     * A File that contains one line per program argument passed to NeoForge for startup. Encoded as UTF-8, unlike
+     * the normal JVM arg-files.
+     * This is additive to {@link #getVanillaProgramArgFile()}.
+     */
+    @OutputFile
+    public abstract RegularFileProperty getNeoForgeProgramArgFile();
+
+    @Inject
+    public InstallProductionClientTask(ExecOperations execOperations) {
+        this.execOperations = execOperations;
+    }
+
+    @TaskAction
+    public void install() throws Exception {
+        var installDir = getInstallDir().getAsFile().get().toPath().toAbsolutePath();
+        Files.createDirectories(installDir);
+
+        // Installer looks for this file
+        var profilesJsonPath = installDir.resolve("launcher_profiles.json");
+        Files.writeString(profilesJsonPath, "{}");
+
+        execOperations.javaexec(spec -> {
+            spec.workingDir(installDir);
+            spec.classpath(getInstaller().getSingleFile());
+            spec.args("--install-client", installDir.toString());
+            try {
+                spec.setStandardOutput(new BufferedOutputStream(Files.newOutputStream(installDir.resolve("install.log"))));
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+
+        var minecraftVersion = getMinecraftVersion().get();
+        var neoForgeVersion = getNeoForgeVersion().get();
+
+        // Download Minecraft Assets and read the asset index id and root for the program arguments
+        var assetPropertiesFile = new File(getTemporaryDir(), "asset.properties");
+        execOperations.javaexec(spec -> {
+            spec.classpath(getNfrt().getSingleFile());
+            spec.args(
+                    "download-assets",
+                    "--minecraft-version",
+                    minecraftVersion,
+                    "--output-properties-to",
+                    assetPropertiesFile.getAbsolutePath()
+            );
+        });
```

Comment:
Instead of using NFRT I would like to see its asset downloading code in a dedicated library. We have many more locations this could be usefull, for example in the installer, and in neogradle.

Additionally downloading assets is a dedicated unique task that is not dependent on NeoForm

Reply:
We can merge this for now, but asset downloading should be extracted from NFRT and moved to a library so that it is reuseable.